hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager/src/main/java/org/apache/hadoop/yarn/server/nodemanager/containermanager/launcher/ContainerLaunch.java

```diff
@@ -170,6 +170,13 @@ public static String expandEnvironment(String var,
     var = var.replace(ApplicationConstants.CLASS_PATH_SEPARATOR,
       File.pathSeparator);
 
+    if (Shell.isJavaVersionAtLeast(17)) {
```

Comment:
Just wondering about this. Wouldn't a node-manager config be easer to reason about?

Implementing like this is a bit strange, currently MR jobs can be configured (default=true) and the distributed shell app sets this regardless of the configuration. During the job submission/config time it's not clear if container will be launched on a Java>=17 node so that's the reason for the placeholder, later ContainerLaunch replaces it to the arg or an empty string. Maybe non MR apps would also require this option to run properly - which they could specify at the job config - but non-homogenous nodes (where nodes have different Java installs) can't be handled easily (maybe with node labels or some other trick).

I think this should be a node-manager config instead. The ContainerLaunch could just simply add the arg when the java version is at least 17 and the config option is set.

Reply:
Had an offline discussion with @tomicooler, the issue with not using a placeholder is the way commands is contructed: in contrast to it's name and type it's a String list containing one element only, the command that is to be executed. In case of Java apps we must add this parameter before the main class, hence I see two options here: add the placeholder right after -Xmx (so that we know it'll be in the correct place) or I can deconstruct the first element of the commands array and add the add-opens flag to the correct place. I think the former is a cleaner, less error-prone solution.